V1.0 to V1.1 docs update

tomczoink opened the pull request:
This adds a new version, v1.1, by moving v1.0 into its own section.

This also includes the capability to auto-detect and render from partials the correct partial for a version without explicitly stating the version (`rest/version/v1.0/_auth` for example can now just be `partial_version('rest/_auth')`). Hopefully this functionality should simplify upkeep of version refs and avoid potential un-updated refs to pages as we get more versions.

This will have a website change to accompany it, which needs to be merged together with this.

mattheworiordan commented:
@tomczoink thanks for this.  Here are some general notes on what is needed to do this release:

- [ ] Diff from v1.0 to v1.1 is updated https://github.com/ably/docs/commit/7d19182e0f42721ac9f4a23cc81f77bf548b5769. We need to tag the release as v1.1 and add release notes (see https://github.com/ably/docs/releases, we will link to a blog post, will get Eve/Kieran/Srushtika to help on that)
- [ ] Need to look at performance issues https://github.com/ably/docs/issues/574
- [ ] Add new methods to the API docs.  Where a method / attribute is not supported by the platform, by design, then we should remove that item completely when that language is selected. However, if a method / attribute should be supported, but is simply not due to missing implementation (see https://docs.google.com/spreadsheets/d/1zzzbAFtTKAnrk0kYyMs55QSHKbtJld1PDF-riRsNleg/edit#gid=0, although need clarification if this is up to date), then we need to introduce a new badge / UI element that indicates a method is not implemented. Let's look at what options we have available that make it clear the method should be available, but are not for that language only, with a link to the README in the repo listing out the missing feature (or perhaps matrix in https://github.com/ably/docs/issues/382)
- [ ] Review changes in v1.0 to v1.1 https://github.com/ably/docs/blob/7d19182e0f42721ac9f4a23cc81f77bf548b5769/content/client-lib-development-guide/versions/features-1-0__1-1.diff, and add method / attribute / object changes to the docs.  Note that most of v1.1 has no impact on the customer-facing APIs, and therefore only internal changes.  Also, there are some new types (for push) and new attributes of types that need to be added.
- [ ] Please focus on Javascript, iOS and Java first for us to review. We can then get help from respective client lib SDK maintainers to give us API details in each language so that we can add.

Note: 

- Idempotent publishing can be ommitted from the docs for now, until it's officially supported in the platform

mattheworiordan commented on content/realtime/channels.textile:
```diff
@@ -255,6 +255,26 @@ channel.publish("action", data: "boom!")
+
+h3(#transient-publish). Transient publishing
+
+When attempting to only publish to a client, it is possible to publish without attaching to the channel. This can be beneficial if you intend to publish into many channels, removing the need to attach to a channel each time you wish to publish.
```
I think we're missing a key aspect of transient publishing being that by not attaching, the client is not subscribing to messages too, so this avoids messages being sent to the client. See https://docs.google.com/document/d/1CSJ3L6Mi2lkw_C26cG_QtdSwcdV8SaiQfEY5MBuENEc/edit#.

mattheworiordan replied:
I am a little concerned that if we do the docs like this, how will we know when to update this as those client SDKs add support?  Perhaps if we go down this route, we should at least add issues in those unsupported repos to update the docs once these features are fully supported. WDYT?

mattheworiordan commented on content/realtime/channels.textile:
```diff
@@ -213,7 +213,7 @@ Ably provides a "REST API":/realtime/channel-metadata to query your app for meta
 h3(#implicit-attach). Implicit attach
 
-Although the attach operation can be initiated explicitly by a client, it is more common for the client to perform a publish or subscribe operation, and the client library will initiate the attach if the channel is not already attached. The client library allows clients to begin publishing "messages":/realtime/messages to a channel as soon as the channel has been created, and messages are queued  ...
+<span lang="default">Although the attach operation can be initiated explicitly by a client, it is more common for the client to perform a publish or subscribe operation, and the client library will initiate the attach if the channel is not already attached. The client library allows clients to begin publishing "messages":/realtime/messages to a channel as soon as the channel has been created, and ...
```
Does Android/Java not support transient publishing?  What about .NET too?

mattheworiordan commented on content/realtime/channels.textile:
```diff
@@ -575,7 +595,7 @@ bq(definition#publish-data).
   objc,swift: publish(name: String?, data: AnyObject?, callback: (("ARTErrorInfo":/realtime/types#error-info?) -> Void)?)
 
-Publish a single message on this channel based on a given event name and payload. A <span lang="default">callback</span><span lang="java">listener</span> may optionally be passed in to this call to be notified of success <span lang="default">or failure</span><span lang="ruby"></span> of the operation. If the channel is @initialized@ (i.e. no attempt to attach has yet been made for this channel),  ...
+Publish a single message on this channel based on a given event name and payload. A <span lang="default">callback</span><span lang="java">listener</span> may optionally be passed in to this call to be notified of success <span lang="default">or failure</span><span lang="ruby"></span> of the operation. <span lang="default">If the channel is @initialized@ (i.e. no attempt to attach has yet been mad ...
```
I realise you've done a great job at keeping track of which libs have implemented these features, but I am concerned that this is going to lead to a lot of issues down the line.  I would almost prefer that we instead document the behaviour by default as this is 1.1 docs, and have a caveat (perhaps a p.note or some unique styling) that indicates this feature is not yet supported in lib X.

tomczoink commented:
For upkeep's sake I've removed the various checks of language, and replaced it with links to downloads, which is where the matrix will go I believe. @mattheworiordan .

mattheworiordan commented on content/realtime/channels.textile:
```diff
@@ -205,15 +205,15 @@ bc[swift]. realtime.channels.get("chatroom").attach { error in
 }
 
-Clients attach to a channel in order to participate on that channel in any way (either to publish, subscribe, or be present on the channel).
+Clients attach to a channel in order to participate on that channel in any way (either to publish, subscribe, or be present on the channel). It is worth noting that some libraries allow for "publishing to a channel without attaching to it":#transient-publish. "See if your chosen SDK supports this":/https://www.ably.io/download.
```
This link is broken, it has a `/` prefix

mattheworiordan commented on content/realtime/channels.textile:
```diff
@@ -584,7 +610,7 @@ bq(definition#publish-msg-array).
   objc,swift: publish(messages: [ "ARTMessage":#message ], callback: (("ARTErrorInfo":/realtime/types#error-info?) -> Void)?)
 
-Publish several messages on this channel. A <span lang="default">callback</span><span lang="java">listener</span><span lang="ruby">block</span> may optionally be passed in to this call to be notified of success <span lang="default">or failure</span><span lang="ruby"></span> of the operation. If the channel is @initialized@ (i.e. no attempt to attach has yet been made for this channel), then calli ...
+Publish several messages on this channel. A <span lang="default">callback</span><span lang="java">listener</span><span lang="ruby">block</span> may optionally be passed in to this call to be notified of success <span lang="default">or failure</span><span lang="ruby"></span> of the operation. <span lang="default">If the channel is @initialized@ (i.e. no attempt to attach has yet been made for this ...
```
Typo, `atach` > `attach`

mattheworiordan commented on content/realtime/channels.textile:
```diff
@@ -805,7 +831,7 @@ bq(definition).
 Attach to this channel ensuring the channel is created in the Ably system and all messages published on the channel will be received by any channel listeners registered using <span lang="default">@subscribe()@</span><span lang="csharp">@Subscribe()@</span><span lang="ruby">@subscribe@</span>. Any resulting channel state change will be emitted to any listeners registered using the "<span lang="def ...
 
-As a convenience, <span lang="default">@attach()@</span><span lang="csharp">@Attach()@</span><span lang="ruby">@attach@</span> will be called implicitly if either "<span lang="default">@publish()@</span><span lang="csharp">@Publish()@</span><span lang="ruby">@publish@</span>":#publish or "<span lang="default">@subscribe()@</span><span lang="csharp">@Subscribe()@</span><span lang="ruby">@subscribe ...
+As a convenience, <span lang="default">@attach()@</span><span lang="csharp">@Attach()@</span><span lang="ruby">@attach@</span> will be called implicitly if either "<span lang="default">@publish()@</span><span lang="csharp">@Publish()@</span><span lang="jsall,ruby,swift"></span>":#publish <span lang="default">or</span> <span lang="jsall,ruby,swift"></span> "<span lang="default">@subscribe()@</span ...
```
This is wrong.  It won't be called implicitly now if publish is called?

mattheworiordan commented on content/rest-api/beta.textile:
````diff
@@ -0,0 +1,187 @@
+
+```[javascript](code-editor:rest/batch-publish)
+var ablyRest = new Ably.Rest({ key: '{{API_KEY}}' })
+ablyRest.request('post', '/messages', null, { "channels": [ "test1", "test2" ], "messages": { "data": 'myData' } }, 
````
Can we wrap this payload or make it a separate var as this won't display in the code editor due to width

tomczoink commented:
Updated per feedback @mattheworiordan . Originally these were 5 different PRs which have been reviewed separately but have now been merged together for the release today, thus the size of this final PR.

I've created a task on asana for myself to go through all the docs once we're deployed.
